src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,781 @@
+    }
+
+    // update function value history
+    for (int i = (int) *(funcValueHistory - 1) - 1; i > 0; --i)
+      funcValueHistory[i] = funcValueHistory[i - 1];
+    funcValueHistory[0] = fitnessValues[index[0]];
+
+    // update xbestever
+    if (xBestEver[N] > population(index[0],N) || gen == 1)
+      for (int i = 0; i <= N; ++i)
+      {
+        xBestEver[i] = population(index[0] , i);
+        xBestEver[N+1] = countevals;
+      }
+
+    const double sqrtmueffdivsigma = std::sqrt(mueff) / sigma;
+    // calculate xmean and rgBDz~N(0,C)
+    for (int i = 0; i < N; ++i)
+    {
+      xold[i] = xmean[i];
+      xmean[i] = 0.;
+      for (int iNk = 0; iNk < mu; ++iNk)
+        xmean[i] += weights[iNk]*population(index[iNk] , i);
+      BDz[i] = sqrtmueffdivsigma*(xmean[i]-xold[i]);
+    }
+
+    // calculate z := D^(-1)* B^(-1)* rgBDz into rgdTmp
+    for (int i = 0; i < N; ++i)
+    {
+      double sum;
+      if (diag)
+        sum = BDz[i];
+      else
+      {
+        sum = 0.;
+        for (int j = 0; j < N; ++j)
+          sum += B(j,i)*BDz[j];
+      }
+      tempRandom[i] = sum/rgD[i];
+    }
+
+    // cumulation for sigma (ps) using B*z
+    const double sqrtFactor = std::sqrt(cs*(double(2)-cs));
+    const double invps = double(1)-cs;
+    for (int i = 0; i < N; ++i)
+    {
+      double sum;
+      if (diag)
+        sum = tempRandom[i];
+      else
+      {
+        sum = double(0);
+        for (int j = 0; j < N; ++j)
+          sum += B(i,j)*tempRandom[j];
+      }
+      ps[i] = invps*ps[i] + sqrtFactor*sum;
+    }
+
+    // calculate norm(ps)^2
+    double psxps(0);
+    for (int i = 0; i < N; ++i)
+    {
+      const double& rgpsi = ps[i];
+      psxps += rgpsi*rgpsi;
+    }
+
+    // cumulation for covariance matrix (pc) using B*D*z~N(0,C)
+    int hsig = std::sqrt(psxps) / std::sqrt(double(1) - std::pow(double(1) - cs, double(2)* gen))
+        / chiN < double(1.4) + double(2) / (N + 1);
+    const double ccumcovinv = 1.-ccumcov;
+    const double hsigFactor = hsig*std::sqrt(ccumcov*(double(2)-ccumcov));
+    for (int i = 0; i < N; ++i)
+      pc[i] = ccumcovinv*pc[i] + hsigFactor*BDz[i];
+
+    // update of C
+    adaptC2(hsig);
+
+    // update of sigma
+    sigma *= std::exp(((std::sqrt(psxps) / chiN) - double(1))* cs / damps);
+
+    state = UPDATED;
+  }
+
+  /**
+   * Some stopping criteria can be set in initials.par, with names starting
+   * with stop... Internal stopping criteria include a maximal condition number
+   * of about 10^15 for the covariance matrix and situations where the numerical
+   * discretisation error in x-space becomes noticeably. You can get a message
+   * that contains the matched stop criteria via getStopMessage().
+   * @return Does any stop criterion match?
+   */
+   template<typename funcType>
+   bool CMAES<funcType>::testForTermination()
+  {
+    double range, fac;
+    int iAchse, iKoo;
+    int diag = diagonalCov == 1 || diagonalCov >= gen;
+
+    std::stringstream message;
```

Comment:
i started finding about this on the standard libraries.. 
thats a nice idea @zoq ..
done